### PR TITLE
Add Qwen Moe 

### PR DESCRIPTION
**Pull request by shivance**

### CODE REVIEW

File: keras_hub/src/models/qwen_moe/qwen_moe_decoder.py [hunk 1]
```diff
@@ -0,0 +1,500 @@
...
+        )
+
+        expert_mask = ops.one_hot(
+            selected_experts, num_classes=self.num_experts
+        )
+        expert_mask = ops.transpose(expert_mask, axes=[2, 1, 0])
+
+        for expert_idx in range(self.num_experts):
+            expert_layer = self.experts[expert_idx]
+            idx, top_x = ops.where(expert_mask[expert_idx])
```
Comment (divyashreepathihalli) [inline keras_hub/src/models/qwen_moe/qwen_moe_decoder.py:174]: will the ops.where return different output shape here on different forward passes? - if so that would not work with JAX XLA

File: keras_hub/src/models/qwen_moe/qwen_moe_decoder.py [hunk 2]
```diff
@@ -0,0 +1,500 @@
...
+        final_hidden_states = ops.zeros(
+            (batch_size * seq_len, hidden_dim), dtype=hidden_states.dtype
+        )
+
+        expert_mask = ops.one_hot(
+            selected_experts, num_classes=self.num_experts
+        )
+        expert_mask = ops.transpose(expert_mask, axes=[2, 1, 0])
+
+        for expert_idx in range(self.num_experts):
```
Comment (divyashreepathihalli) [inline keras_hub/src/models/qwen_moe/qwen_moe_decoder.py:172]: The for loop here to go over the experts are inefficient for XLA compilation. This implementation would need to updated - I had tried out a dummy MOE implementation in JAX here - https://colab.sandbox.google.com/drive/1r0rscZK_2bNpDmFLC1POEEQoKcqWQYlQ
I order to bring this to KearsHub we are missing ragged_dot op.

**Review by divyashreepathihalli — COMMENTED** [review]
just reviewed the MOE part of the code.

File: keras_hub/src/models/qwen_moe/qwen_moe_decoder.py (hunk 2 above)
Comment (divyashreepathihalli) [reply keras_hub/src/models/qwen_moe/qwen_moe_decoder.py:172]: I have prototyped the implementation - will add this op soon

File: keras_hub/src/models/qwen/qwen_decoder.py [hunk 3]
```diff
@@ -79,7 +79,7 @@ def build(self, decoder_sequence_shape):
         self.hidden_dim = decoder_sequence_shape[-1]
 
         # Self attention layer.
-        self._self_attention_layer = QwenAttention(
+        self._self_attention_layer = QwenMoeAttention(
```
[the remark this answers is not in the document]
Comment (shivance) [reply keras_hub/src/models/qwen/qwen_decoder.py]: looks like this slipped in in find & replace, fixed it.

File: keras_hub/src/models/qwen_moe/qwen_moe_decoder.py (hunk 1 above)
Comment (shivance) [reply keras_hub/src/models/qwen_moe/qwen_moe_decoder.py:174]: I don't think that this should be the case. Why do you think so?

File: keras_hub/src/models/qwen_moe/qwen_moe_decoder.py (hunk 2 above)
Comment (shivance) [reply keras_hub/src/models/qwen_moe/qwen_moe_decoder.py:172]: When can we expect this to be available as a part of keras op?

**Comment by shivance** [comment]
@divyashreepathihalli How should we accomodate aux_loss for CausalLM task here model here?

We are specifying Sparse Categorical CrossEntropy Loss here:

https://github.com/keras-team/keras-hub/blob/b9974443ca751546932c09955dfe5ccac99164e3/keras_hub/src/models/causal_lm.py#L109-L119

File: keras_hub/src/models/qwen_moe/qwen_moe_decoder.py (hunk 1 above)
Comment (abheesht17) [reply keras_hub/src/models/qwen_moe/qwen_moe_decoder.py:174]: Whether the output shape is differing or not doesn't matter. `ops.where` won't work on JAX at all, if you don't provide x and y.

`ops.where` calls [`jnp.where`](https://docs.jax.dev/en/latest/_autosummary/jax.numpy.where.html). Check the note [here](https://docs.jax.dev/en/latest/_autosummary/jax.numpy.nonzero.html#jax.numpy.nonzero):

```
Because the size of the output of nonzero is data-dependent, the function is not compatible with JIT and other transformations. The JAX version adds the optional size argument which must be specified statically for jnp.nonzero to be used within JAX’s transformations.
```

**Review by divyashreepathihalli — COMMENTED** [review]
Thanks for the updates @heyyanshuman!
I left some comments on the PR regarding tf ops
please add tests for the layers, backbones and tasks
I am curious to know if model.fit works, do you have a demo colab for inference and FT? - looking for the aux loss implementation

File: keras_hub/src/models/qwen_moe/qwen_moe_attention.py [hunk 4]
```diff
@@ -0,0 +1,360 @@
...
+            output_shape=(None, self.num_query_heads, head_dim),
+            kernel_initializer=self.kernel_initializer,
+            bias_initializer=self.bias_initializer,
+            bias_axes="uh",
+            dtype=self.dtype_policy,
+            name="query",
+        )
+        self._query_dense.build(inputs_shape)
+
+        self._key_dense = keras.layers.EinsumDense(
```
[the remark this answers is not in the document]
Comment (heyyanshuman) [reply keras_hub/src/models/qwen_moe/qwen_moe_attention.py:90]: I don't have access to this document :(

File: keras_hub/src/models/qwen_moe/qwen_moe_attention.py [hunk 5]
```diff
@@ -0,0 +1,360 @@
...
+            query: Query tensor.
+            key: Key tensor.
+            value: Value tensor.
+            attention_mask: Optional mask tensor.
+            cache_update_index: Index for sliding window computation.
+
+        Returns:
+            attention_output: Output tensor after applying attention.
+        """
+        if fused_attention_op_available():
```
[the remark this answers is not in the document]
Comment (heyyanshuman) [reply keras_hub/src/models/qwen_moe/qwen_moe_attention.py:266]: I see, replicated the change here.

File: keras_hub/src/models/qwen_moe/qwen_moe_attention.py [hunk 6]
```diff
@@ -0,0 +1,360 @@
...
+        # Compute the sliding window for square attention.
+        all_ones = ops.ones((key_len, key_len), "bool")
+        if keras.config.backend() == "tensorflow":
+            # TODO: trui/tril has issues with dynamic shape on the tensorflow
+            # backend. We should fix, but use `band_part` for now.
+            import tensorflow as tf
+
+            band_size = ops.minimum(key_len, self.sliding_window_size - 1)
+            band_size = ops.cast(band_size, "int32")
+            sliding_mask = tf.linalg.band_part(all_ones, band_size, band_size)
```
[the remark this answers is not in the document]
Comment (heyyanshuman) [reply keras_hub/src/models/qwen_moe/qwen_moe_attention.py:328]: copyover from gemma attention, removed

File: keras_hub/src/models/qwen_moe/qwen_moe_causal_lm.py [hunk 7]
```diff
@@ -0,0 +1,297 @@
...
+            ["This is a", "Where are you"],
+            max_length=30
+        )
+        preprocessed = qwen_lm.preprocessor.generate_preprocess(generations)
+        generation_ids = preprocessed["token_ids"]
+        padding_mask = preprocessed["padding_mask"]
+        target_ids = keras.ops.roll(generation_ids, shift=-1, axis=1)
+
+        embeddings = None
+        with tf.GradientTape(watch_accessed_variables=True) as tape:
```
[the remark this answers is not in the document]
Comment (heyyanshuman) [reply keras_hub/src/models/qwen_moe/qwen_moe_causal_lm.py:237]: copiedover docstrings from llama (i guess), removed

File: keras_hub/src/models/qwen_moe/qwen_moe_attention.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (divyashreepathihalli) [reply keras_hub/src/models/qwen_moe/qwen_moe_attention.py:94]: sorry! wrong copy pasta error! updated the link - https://github.com/keras-team/keras-hub/blob/7c8694212f1b1c583d927631536d5e15b4a133f9/keras_hub/src/models/backbone.py#L195

File: keras_hub/src/models/qwen_moe/qwen_moe_attention.py [hunk 8]
```diff
@@ -0,0 +1,377 @@
...
+        return self._softmax(attention_scores)
+
+    def _use_fused_attention_op(self):
+        if not fused_attention_op_available():
+            return False
+        if self.dropout > 0.0:
+            return False
+        if running_on_gpu():
+            # GPU never supports softcap in the fused op.
+            if self.logit_soft_cap is not None:
```
Comment (divyashreepathihalli) [inline keras_hub/src/models/qwen_moe/qwen_moe_attention.py:260]: does Qwen MOE use logit_soft_cap?

**Review by divyashreepathihalli — COMMENTED** [review]
Thanks @heyyanshuman  - can you add a demo colab for inference and fit? and also aprovide colab/screenshot for numerics verification?

File: keras_hub/src/models/qwen_moe/qwen_moe_decoder.py [hunk 9]
```diff
@@ -0,0 +1,636 @@
...
+        )
+
+        self.activation = keras.activations.get(self.activation_fn)
+        self.built = True
+
+    def call(self, x):
+        gate_output = self._feedforward_gate_dense(x)
+
+        # Note that we run the activation function in full 32-bit
+        # precision since this is what `torch.nn.functional.silu`
```
Comment (divyashreepathihalli) [inline keras_hub/src/models/qwen_moe/qwen_moe_decoder.py:147]: is this torch specific? Or did you mean this is what the original implementation is doing?

File: keras_hub/src/models/qwen_moe/qwen_moe_decoder.py [hunk 10]
```diff
@@ -0,0 +1,636 @@
...
+
+        x = self._feedforward_intermediate_dense(x)
+
+        x = self._feedforward_output_dense(ops.multiply(x, gate_output))
+
+        return x
+
+
+class QwenMoeExperts(keras.layers.Layer):
+    """Batched feed‑forward experts à‑la Llama‑4 (pure keras.ops)."""
```
Comment (divyashreepathihalli) [inline keras_hub/src/models/qwen_moe/qwen_moe_decoder.py:165]: NIT : `q-la` - lets keep this simple english  - inspired by Llama 4 or whatever

File: keras_hub/src/models/qwen_moe/qwen_moe_decoder.py [hunk 11]
```diff
@@ -0,0 +1,636 @@
...
+        gate, up = ops.split(gate_up, 2, axis=-1)
+        hidden = up * self.activation(gate)
+        out = ops.einsum(
+            "eti,eih->eth", hidden, self._expert_feedforward_output_dense
+        )
+        return out
+
+
+class QwenSparseMoeBlock(keras.layers.Layer):
+    """Qwen‑2 sparse block rewritten in Llama‑4 batched style."""
```
Comment (divyashreepathihalli) [inline keras_hub/src/models/qwen_moe/qwen_moe_decoder.py:219]: NIT: rewritten - inspired by Llama 4 implementation

File: keras_hub/src/models/qwen_moe/qwen_moe_decoder.py [hunk 12]
```diff
@@ -0,0 +1,636 @@
...
+        self,
+        hidden_dim,
+        moe_intermediate_dim,
+        shared_expert_intermediate_dim,
+        num_experts,
+        top_k,
+        norm_topk_prob,
+        kernel_initializer="glorot_uniform",
+        layer_norm_epsilon=1e-5,
+        router_aux_loss_coef=0.01,
```
Comment (divyashreepathihalli) [inline keras_hub/src/models/qwen_moe/qwen_moe_decoder.py:231]: expand coef-> coefficient

File: keras_hub/src/models/qwen_moe/qwen_moe_decoder.py [hunk 13]
```diff
@@ -0,0 +1,636 @@
...
+        **kwargs,
+    ):
+        super().__init__(**kwargs)
+        self.hidden_dim = hidden_dim
+        self.intermediate_dim = moe_intermediate_dim
+        self.intermediate_dim_shared = shared_expert_intermediate_dim
+        self.num_experts = num_experts
+        self.top_k = top_k
+        self.norm_topk_prob = norm_topk_prob
```
Comment (divyashreepathihalli) [inline keras_hub/src/models/qwen_moe/qwen_moe_decoder.py:240]: topk->top_k

File: keras_hub/src/models/qwen_moe/qwen_moe_decoder.py [hunk 14]
```diff
@@ -0,0 +1,636 @@
...
+    def __init__(
+        self,
+        intermediate_dim,
+        num_query_heads,
+        num_key_value_heads,
+        moe_intermediate_dim,
+        shared_expert_intermediate_dim,
+        num_experts,
+        top_k,
+        norm_topk_prob,
```
Comment (divyashreepathihalli) [inline keras_hub/src/models/qwen_moe/qwen_moe_decoder.py:347]: topk->top_k - here and everywhere

File: keras_hub/src/models/qwen_moe/qwen_moe_decoder.py [hunk 15]
```diff
@@ -0,0 +1,636 @@
...
+        mlp_only_layers=[],
+        output_router_logits=False,
+        router_aux_loss_coef=0.001,
+        **kwargs,
+    ):
+        super().__init__(**kwargs)
+        self.intermediate_dim = intermediate_dim
+        self.num_query_heads = num_query_heads
+        self.num_key_value_heads = num_key_value_heads
+
```
Comment (divyashreepathihalli) [inline keras_hub/src/models/qwen_moe/qwen_moe_decoder.py:367]: remove unnecessary new lines

File: keras_hub/src/models/qwen_moe/qwen_moe_backbone_test.py [hunk 16]
```diff
@@ -0,0 +1,98 @@
+import pytest
+from keras import ops
+
+from keras_hub.src.models.qwen_moe.qwen_moe_backbone import QwenMoeBackbone
+from keras_hub.src.tests.test_case import TestCase
+
+
+class QwenMoeBackboneTest(TestCase):
+    def setUp(self):
```
Comment (divyashreepathihalli) [inline keras_hub/src/models/qwen_moe/qwen_moe_backbone_test.py:9]: add a flash attention mock test like - https://github.com/keras-team/keras-hub/blob/a5337f5247cc15241290c5e2e44a23dd15577e5e/keras_hub/src/models/gemma/gemma_causal_lm_test.py#L100

File: keras_hub/src/models/qwen_moe/qwen_moe_decoder.py [hunk 17]
```diff
@@ -0,0 +1,636 @@
...
+                num_experts=self.num_experts,
+                top_k=self.top_k,
+                attention_mask=attention_mask,
+            )
+            self.add_loss(self.router_aux_loss_coef * aux_loss)
+
+        return out, router_logits
+
+
+class QwenMoeTransformerDecoder(keras.layers.Layer):
```
Comment (divyashreepathihalli) [inline keras_hub/src/models/qwen_moe/qwen_moe_decoder.py:337]: add a self.run_layer_test for the decoder

File: keras_hub/src/models/qwen_moe/qwen_moe_decoder.py [hunk 18]
```diff
@@ -0,0 +1,636 @@
...
+        gate_output = ops.cast(gate_output, self.compute_dtype)
+
+        x = self._feedforward_intermediate_dense(x)
+
+        x = self._feedforward_output_dense(ops.multiply(x, gate_output))
+
+        return x
+
+
+class QwenMoeExperts(keras.layers.Layer):
```
Comment (divyashreepathihalli) [inline keras_hub/src/models/qwen_moe/qwen_moe_decoder.py:164]: also add a self.run_layer_test for this layer as well

**Review by divyashreepathihalli — COMMENTED** [review]
Left a few more NIT comments!! Looking great overall!
Once we have the comments addressed and inference and fit demo - it is ready for merge.

File: keras_hub/src/models/qwen_moe/qwen_moe_decoder.py (hunk 17 above)
Comment (heyyanshuman) [reply keras_hub/src/models/qwen_moe/qwen_moe_decoder.py:337]: i just noticed that we don't have layer tests for other model? Why is that the case?

Comment (divyashreepathihalli) [reply keras_hub/src/models/qwen_moe/qwen_moe_decoder.py:337]: Ahh, sorry only needed for exported layers - no need to add them for these

**Comment by kanpuriyanawab** [comment]
Qwen moe output matching:

Screenshot 2025-04-20 at 2 18 58 PM

File: tools/checkpoint_conversion/convert_qwen_moe_checkpoints.py [hunk 19]
```diff
@@ -0,0 +1,162 @@
...
+    )
+    hf_generated_text = hf_tokenizer.batch_decode(
+        outputs, skip_special_tokens=True
+    )[0]
+    print("🔶 Huggingface output:", hf_generated_text)
+
+
+def main(_):
+    # === Get the preset name ===
+    # if FLAGS.preset not in PRESET_MAP.keys():
```
Comment (divyashreepathihalli) [inline tools/checkpoint_conversion/convert_qwen_moe_checkpoints.py:123]: uncomment the code

File: tools/checkpoint_conversion/convert_qwen_moe_checkpoints.py [hunk 20]
```diff
@@ -0,0 +1,162 @@
...
+    #     raise ValueError(
+    #         f"Invalid preset {FLAGS.preset}. Must be one "
+    #         f"of {','.join(PRESET_MAP.keys())}"
+    #     )
+    # preset = FLAGS.preset
+    # hf_preset = PRESET_MAP[preset]
+    hf_preset = "Qwen/Qwen1.5-MoE-A2.7B"
```
Comment (divyashreepathihalli) [inline tools/checkpoint_conversion/convert_qwen_moe_checkpoints.py:130]: let us not hardcode this

File: keras_hub/src/models/qwen_moe/qwen_moe_decoder.py [hunk 21]
```diff
@@ -0,0 +1,625 @@
...
+
+        # Compute and add auxiliary loss during training
+        if training:
+            aux_loss = compute_load_balancing_loss(
+                router_logits=router_logits,
+                num_experts=self.num_experts,
+                top_k=self.top_k,
+                attention_mask=attention_mask,
+            )
+            self.add_loss(self.router_aux_loss_coefficient * aux_loss)
```
Comment (divyashreepathihalli) [inline keras_hub/src/models/qwen_moe/qwen_moe_decoder.py:332]: nice!!

**Review by divyashreepathihalli — COMMENTED** [review]
Thanks Anshuman! left a few NIT comments. Can you please add a presets file as well?

**Comment by divyashreepathihalli** [comment]
Also I want to see Generate output matching

File: tools/checkpoint_conversion/convert_qwen_moe_checkpoints.py (hunk 19 above)
Comment (kanpuriyanawab) [reply tools/checkpoint_conversion/convert_qwen_moe_checkpoints.py:123]: my bad, it was for local testing

**Comment by kanpuriyanawab** [comment]
generate output matching

**Review by divyashreepathihalli — APPROVED** [review]
